### ISSUE TITLE
SourceCred: measuring who built a project

### ISSUE THREAD

**Issue by themightychris** [sfosc/sfosc#61]
[SourceCred](https://sourcecred.io/) is an interesting project that might be relevant to some of the conversations here, it aims to help address OSS sustainability by creating a way to quantify the work put into an open source project by each contributor (in a deeper way than just commits/lines-of-code). They've been supported by grants from [Protocol Labs](https://protocol.ai/) who I know is keen on a lot of the high-level goals discussed here

What relevance do folks here see in quantifying contributions to the sustainability of projects? Should it matter for governance? For business models? Just for research/curiosity sake?

**Comment by Beanow** [comment]
Definitely looks interesting to me. I think it has relevance for sure.
At the very least it could be used to illustrate the problem. Bus factor, burnout risk, etc. Has some similarities with how https://chaoss.community/ quantifies project health.

It seems to focus on the problem of fair distribution of benefits. Whether that is influence on governance, currency or others. Projects running with this would be interesting to watch as an experiment in finding out what fairness is. And the software would be interesting to use when you have a considerable stream of benefits you need to distribute more fairly.

On the other hand it should definitely be part of a larger discussion / toolkit / strategy, because applied as-is to your average donation backed project, I fear it would only make fighting over crumbs more transparent.

Should we index sustainability related tools?

**Comment by decentralion** [comment]
Hi there, 👋 I'm the maintainer of SourceCred. I'm reading through the SFOSC book with interest. @Beanow I think you hit the nail on the head in that SourceCred could fit pretty well into the governance as well as distribution of benefits. The SFOSC [chapter on governance](https://sfosc.org/sfosc-book/governance/) discusses a bit how there need to be impartial systems for validating peoples contributions, SourceCred could play a role there. In the future, I also hope we can use cred to build new economic models for open-source communities. 

I also imagine that SourceCred will be useful for managing projects. For example, imagine a project that has way too few people doing documentation work. Rather than having a hierarchical structure where someone tells other people to work on documentation, the community could increase the "cred weight" on documentation, so that all contributors are incentivized to put more effort into the docs side.

I recently wrote [an essay](https://medium.com/sourcecred/introduction-to-sourcecred-7665297af715) exploring SourceCred as a tool for organizing creative communities a bit more, you may find it interesting.

Thanks to @nothingismagick, I found SFOSC (and this thread) because he dropped by the SourceCred [discord server](http://sourcecred.io/discord-invite). :)

**Comment by Beanow** [comment]
The bulk of SFOSC work happens here on GitHub. Maybe we can use SourceCred for attribution? Since very helpful discussions happen in issues/PRs but aren't necessarily acknowledged accordingly by GitHub.

As well as the suggestion
> For example, imagine a project that has way too few people doing documentation work.

Right now we're just switching to #71 two reviews being required for merges and have a risk of stagnating PRs. Increasing weights on reviews seems appropriate here.

Would either of you be able to help configure a setup for SFOSC?

**Comment by decentralion** [comment]
I set up a demo instance for you [here](https://decentralion.github.io/sfosc-sourcecred-demo/site/prototype/sfosc/sfosc/). Going to go through some of the interesting results and features.

Here's the top level cred distribution:
![image](https://user-images.githubusercontent.com/1400023/58102209-94780200-7be9-11e9-96ad-63bd7032e699.png)

It's cool to see such an even distribution of cred, it shows that there's a lot of people involved in SFOSC, not just one or two core contributors. 

As you mentioned, it's possible to change the weight on reviews. With the default settings @Beanow you get ~5% of your cred from reviews:
![image](https://user-images.githubusercontent.com/1400023/58103071-0270f900-7beb-11e9-9e4c-c71cb2e78276.png)

But if we change the weight on reviews from 1x to 4x (the same as pull requests) then it goes up to ~20%:
![image](https://user-images.githubusercontent.com/1400023/58102459-fd5f7a00-7be9-11e9-845c-1a8b0130bbb4.png)

Interestingly, the overall distribution of cred stays similar as we increase the weight on reviews, which is  because the top three contributors (@adamhjk, @Beanow, and @nothingismagick) are also the only three contributors who have written reviews. Increasing the weight on reviews might be a good way to signal to other people that their feedback on reviews would be valued. :smile: 

To play around with changing the type-level weights yourself, just open the weight configuration, change the sliders, and press the "analyze cred" button again.

Another feature we [recently added](https://github.com/sourcecred/sourcecred/pull/1143) is the ability to set custom weights on individual nodes. In the case of SFOSC, since there is a relatively small number of issues and pull requests, you could set a weight on each one based on how much value you think it brought to the project. This could improve the quality of cred a lot if some issues/pulls were a lot more important than others. The weights can be downloaded/uploaded via the little icons next to the weight configuration button.

![image](https://user-images.githubusercontent.com/1400023/58102973-d9506880-7bea-11e9-836f-71d9a6768ab4.png)

The instance I linked above is static and hosted out of my GitHub pages. Let me know if SFOSC is interested in running its own instance, we can work together to host it out of a SFOSC repo's github pages, and configure it to suit the needs of this project.

**Comment by Beanow** [comment]
Thank you :D very cool stuff and insightful breakdown of first results.

We do have github pages powering https://sfosc.org/
Which is hugo built from this repo, committing output into https://github.com/sfosc/sfosc.github.io
If we can somehow integrate the two as more prominent :heart: for the community I think that would be good.

Additionally I think trying to converge on a fair set of weights sounds like a great meta discussion to have!

Any recommendations on how to best do a deployment / integration?

**Comment by nothingismagick** [comment]
Yeah, I think its really promising too. Love it!

**Comment by themightychris** [comment]
@decentralion this is awesome! Thanks for setting up a demo

One thing it's got me thinking: have you considered making it possible to treat a set of repositories as one? Many projects mights split up their components across several repositories instead of employing a monorepo pattern, but it might be most helpful to analyze such a set of repositories as if they were one

**Comment by decentralion** [comment]
> Any recommendations on how to best do a deployment / integration?

The simplest approach would be to put the cred attribution in its own GitHub repository, along with an updater script which, when run, updates the distribution. That repository would have its own url via github pages, and you could link to it from the main site. The downside is it won't be well integrated into your website--it will look and feel like a separate page. If you want to take this approach, I can help set up the repository (maybe create a new repo for it and I'll send a pull request setting it up). 

We could imagine integrating it more deeply, e.g. exporting the React component for the cred explorer and then embedding it directly in your site. I expect this would be a lot of work though, so I think we should hold off on it for now.

> One thing it's got me thinking: have you considered making it possible to treat a set of repositories as one? Many projects mights split up their components across several repositories instead of employing a monorepo pattern, but it might be most helpful to analyze such a set of repositories as if they were one. Or are there reasons it needs to be anchored to one repo?

Yup, this is totally on the roadmap. It's experimentally supported right now; there's a [branch](https://github.com/sourcecred/sourcecred/tree/org-score-prototype#sourcecred---org-scores-sidebranch) that allows you to load whole organizations rather than individual repos. Unfortunately, it's kind of hacky--SourceCred continues to operate on individual repositories, so we first aggregate all the different repositories into one "super repository" and then compute its cred. Not very elegant, and there's a nasty [bug](https://github.com/sourcecred/sourcecred/issues/904) which causes numeric GitHub references (e.g. `See #904`) to get misdirected to point to issues and pulls in the wrong repo.

Ultimately I want to change the way SourceCred identifies cred instances to be organized around [projects](https://github.com/sourcecred/sourcecred/issues/701) rather than individual repositories, and then each project can have pointers to many repositories, and also other sources of data, like Discourse forums, Medium accounts, etc. Just haven't gotten around to it yet. Contributions welcome :smile:

**Comment by Beanow** [comment]
@decentralion yes I think a styling integration is a bit overkill for now. Priorities for a deployment now I think should be easy upgrade path and security. The readme mentions personal API tokens, though I think they are a little too coarse and risky for automation. Are there alternative auth schemes available or should that be scoped with a new github user?

Atm we don't have full access to set everything up either so an experiment over on a seperate repo to link to from the readme would be a good start. (Organization owner is on leave, helping to maintain in the mean time)

**Comment by decentralion** [comment]
Regarding the API token; it doesn't need any user scoped permissions (e.g. it doesn't need access to read a user's own repositories), so creating a dummy GitHub account like sfosc-credhelper and using a token from it would be fine.

If you don't have access to create a new repo yet: how about we just use the instance I set up (https://decentralion.github.io/sfosc-sourcecred-demo/site/prototype/sfosc/sfosc/) for now? It's hosted here: https://github.com/decentralion/sfosc-sourcecred-demo, and I can add an updater script to the repo to make it easier to maintain, and add you (and whomever you like) as maintainers of that repo. Then, once the sfosc owner is back, we can transfer that repo to the sfosc org and rename it as is appropriate.

**Comment by Beanow** [comment]
@decentralion As we've experimented with repos in the org we can create right now. A sub-directory will work. Like https://sfosc.org/preview/ which is in https://github.com/sfosc/preview, but lacking org ownership it can't be connected with CI platforms easily right now. With that in mind I think starting from your demo repo is best :]

It would be great if you can add an updater script. If there's anything you need me to do to get it running let me know. Maintainers I would like to suggest for the repository right now: @adamhjk, @nothingismagick and myself.

**Comment by Beanow** [comment]
Although it includes a couple of workarounds and changes to sourcecred from master...
There's a more official instance now set up!

https://github.com/sfosc/sourcecred should update on a daily cron.
Can be tested over at https://sfosc.org/sourcecred/prototype.

Anyone interested in tinkering with the default weights, you can make pull requests for [`weights.toml`](https://github.com/sfosc/sourcecred/blob/master/weights.toml).
